src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingGuards.java

```diff
@@ -454,7 +470,38 @@ public PlayerEntity getPlayerToFollowOrRally()
     @Override
     public void onColonyTick(@NotNull final IColony colony)
     {
-        super.onColonyTick(colony);
+
+        // If we have no active worker, attempt to grab one from the appropriate trainer
+        if (hireTrainees && !isFull() && ((getBuildingLevel() > 0 && isBuilt()))
+              && (this.getHiringMode() == HiringMode.DEFAULT && !this.getColony().isManualHiring() || this.getHiringMode() == HiringMode.AUTO))
+        {
+            List<ICitizenData> trainingCitizens = null;
+            
+            if(this.getGuardType() == ModGuardTypes.ranger)
+            {
+                trainingCitizens = colony.getCitizenManager().getCitizens().stream().filter(x -> x.getJob() instanceof JobArcherTraining).collect(Collectors.toList());
+            }
+            else if (this.getGuardType() == ModGuardTypes.knight)
+            {
+                trainingCitizens = colony.getCitizenManager().getCitizens().stream().filter(x -> x.getJob() instanceof JobCombatTraining).collect(Collectors.toList());
+            }
+
+            if(trainingCitizens != null && !trainingCitizens.isEmpty())
+            {
+                Comparator<ICitizenData> skillComparator = Comparator.comparingInt(e -> e.getCitizenSkillHandler().getLevel(job.getPrimarySkill()));
```

Comment:
actually you can do the whole find of job + find best skill logic with one single loop instead, that should be a lot better performance wise, since you only need to find the best skilled one

Reply:
Yup. Cleaned up.